Add onboarding payment gateway setup methods

joshuatf opened the pull request:
# Changes proposed in this Pull Request:

Part of - woocommerce/woocommerce#32258

Adds the methods needed to use the config-based setup for the payment gateway during onboarding.

<img width="707" alt="Screen Shot 2021-05-27 at 11 31 15 AM" src="https://user-images.githubusercontent.com/10561050/119854694-49cba780-bedf-11eb-99f2-55d627b301b9.png">
<img width="705" alt="Screen Shot 2021-05-27 at 11 31 11 AM" src="https://user-images.githubusercontent.com/10561050/119854697-4a643e00-bedf-11eb-845d-28870d3ba3e9.png">


# Testing instructions

1. Install and activate this plugin in your `plugins/` directory matching the wp.org slug.  `woocommerce-gateway-stripe`
2. Install and activate the development version of [WooCommerce Admin](https://github.com/woocommerce/woocommerce-admin) ( `git clone https://github.com/woocommerce/woocommerce-admin.git && cd woocommerce-admin && composer install && npm i && npm run dev` ).
3. Make sure a qualifying Stripe country is selected for the store base location in WooCommerce general settings and that the "cbd" industry is not selected during onboarding.
2. Navigate to the Stripe gateway in the payments task `/wp-admin/admin.php?page=wc-admin&task=payments`
3. Fill out the required setting fields.
4. Make sure that the Stripe gateway is moved to the "Enabled payment gateways" section.
5. Make sure that settings are persisted on page refresh.
6. When adding testing and publishable keys that start with `test_`, test mode should be toggled on.  Currently this does not mark setup as "complete" but I think this open to interpretation and we can update accordingly.

joshuatf committed:
Add remote payment gateway setup methods

v18 commented:
👋  Hey @joshuatf - I see this PR is for testing purposes. I'm not sure if there's anything we can help with -- if you do need any assistance with anything, let us know in the Fractal team channel.

joshuatf committed:
Add needs_setup method

joshuatf committed:
Add missing return param

joshuatf committed:
Filter the onboarding settings depending on types of keys entered

joshuatf committed:
Use get_option to get latest updated setting values

joshuatf commented:
Hey @v18 👋 

We've finalized our API for this in WooCommerce Admin so we're planning to go forward with this PR.  I've updated the PR with testing instructions.  Please let me know if there's anything that's not clear or I'm also happy to catch up in Slack and discuss further.

joshuatf committed:
Update connection URL method name

asumaran commented:
Not sure I'm testing this correctly. UI looks slightly different to me. Also, installing `WooCommerce Admin` makes no difference. Am I missing something?. Here's the [installed plugins](https://user-images.githubusercontent.com/1025173/121819477-e10c5b00-cc52-11eb-8977-7d857eee7988.png) I'm using.

<img width="711" alt="Screen Shot 2021-06-13 at 2 21 35 PM" src="https://user-images.githubusercontent.com/1025173/121819428-a99dae80-cc52-11eb-80df-a70e7e80d98f.png">

I'm going to unassign myself so someone more familiar with this can take a look at this next week.

joshuatf commented:
Thanks for taking a look, @asumaran.  We can wait for someone else to take a look at this.

I think the issue was the built feature config.  I modified the above testing and build instructions for WooCommerce Admin slightly so that the new feature is turned on.

dechov commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -1311,4 +1312,25 @@ public function get_setup_help_text() {
+	 */
+	public function update_onboarding_settings( $settings ) {
+		if ( ! empty( $settings['publishable_key'] ) && ! empty( $settings['secret_key'] ) ) {
+			if ( strpos( $settings['publishable_key'], 'test_' ) === 0 || strpos( $settings['secret_key'], 'test_' ) === 0 ) {
```
The keys start with `pk_` and  `sk_` respectively, so this check will need to include those (or be modified in some other way).

joshuatf replied:
Is this true for the test keys?  I seem to recall the logic in WooCommerce Admin checking for `test_` prefixes if they were test keys.

dechov replied:
Yup, the test keys are prefixed with `pk_test_` and `sk_test_`

joshuatf replied:
Sorry, I see this was `sk_test` but also updated in https://github.com/woocommerce/woocommerce-admin/pull/5353.

I'll update accordingly once we have a decision on the test mode issue below.

dechov commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -1266,4 +1267,70 @@ public function settings_api_sanitized_fields( $settings ) {
+				unset( $settings['secret_key'] );
+				$settings['testmode'] = 'yes';
+			} else {
+				$settings['testmode'] = 'no';
```
Would this not turn off test mode any time the gateway is updated after live keys are set (via REST API, which looks like the only time this filter is run)?

joshuatf replied:
It does.  Can both live keys and test keys exist simultaneously?

This would be much better if we could check if the request was being made from WooCommerce Admin, but AFAIK the only params allowed are the predefined settings in the gateway.

dechov replied:
Sorry, didn't see this comment before replying below – yup, live and test keys can both be set, and it's just the `testmode` setting that switches between them.

dechov reviewed (changes requested):
Looks good – noted a couple of things below.

I see that the input captions are "Live Publishable Key" and "Live Secret Key", which I assume are grabbed from the settings. IIRC previously via WC Admin, it didn't say "Live" since it works with both live and test keys. I don't think it's a big deal though – just pointing it out.

_Edit:_ needs a quick rebase as well, just placing these methods after a newly added one.

joshuatf commented:
Thanks for the review @dechov!

>I see that the input captions are "Live Publishable Key" and "Live Secret Key", which I assume are grabbed from the settings.

That's correct.  We could simplify this PR and just allow live keys.  The turning off of test mode was just to mimic previous behavior, but I'm open to suggestions on this.

dechov commented:
👍 Personally I'm a fan of allowing test keys (see https://github.com/woocommerce/woocommerce-admin/issues/4391, https://github.com/woocommerce/woocommerce-admin/pull/5201, and https://github.com/woocommerce/woocommerce-admin/pull/5353), but my [comment above](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1569#discussion_r651267830) refers to the code looking to me like it turns off test mode in cases when it isn't intended to be turned off.

joshuatf committed:
Only toggle test mode based on keys when referer is wc-admin

joshuatf commented:
@dechov Updated the logic in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1569/commits/7c18ac61478024703468a87eed669a4886369e76.  The logic to toggle test mode now only occurs when the request was made from the `wc-admin` page.  If test keys are provided, test mode is turned on; with any other keys the test mode is turned off.

Can you let me know what you think of this?

joshuatf commented:
@dechov Could you take a look at this again when you have time?  This would be helpful to increase conversion of this flow in payment setup.

lovo-h commented:
@joshuatf, @dechov, I'll pick this up and take a look 👀

lovo-h commented:
@joshuatf, I will go ahead and rebase off of `develop` (see [this comment](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1569#pullrequestreview-683306867)). Confirmed with you through Slack 👍

lovo-h commented:
✅  Short-circuiting the logic (to toggle test mode) if the page is not `wc-admin` or the task is not `payments` solves the issue where `test mode` would turn off any time the gateway was updated.
❌  The testing instructions might need to be updated. See the **Testing instructions** below for more details.

#### Testing instructions

1. Install and activate this plugin (`woocommerce-gateway-stripe`).
2. Install the development version of [WooCommerce Admin plugin](https://github.com/woocommerce/woocommerce-admin) by running the following terminal commands:
```bash
git clone https://github.com/woocommerce/woocommerce-admin.git \
&& cd woocommerce-admin \
&& composer install \
&& npm i \
&& npm run dev
```
3. Login into WordPress on an unsecured connection (HTTP) in order to obtain the following form (see image below):

![image](https://user-images.githubusercontent.com/6844516/142940871-bba0fc3a-3a6c-4ff4-a035-6042d297df59.png)

**Note**: Otherwise, if one logs into WordPress through a secured connection (HTTPS), the following form is displayed (see image below):

![image](https://user-images.githubusercontent.com/6844516/142941125-9969d01f-1716-44df-b5c6-c9b8f4fffd1c.png)

4. Go to `wc-admin > plugins` and `Activate` the [WooCommerce Admin plugin](https://github.com/woocommerce/woocommerce-admin).
5. Click on the `WooCommerce` menu-item and fill out the following sections:
* Store Details (Make sure a qualifying Stripe country is selected)
* Industry (Make sure not to select the "CBD" industry)
* Product Types
* Business Details
* Theme

![image](https://user-images.githubusercontent.com/6844516/142947458-f6c7807f-f08e-42a4-bc85-311a81f02a8d.png)

6. Click `Set up payments` and then, in the Stripe box, click the `Finish setup` button.
7. Fill out the required setting fields (enter live/test keys), click `Proceed`, then click `Set up payments`.

**Note**: Here, I wanted to highlight that I was not entirely sure if the `Proceed` button should have automatically navigated to the next step.

8. Go to `WooCommerce > Settings > Payments` and ensure `Stripe` is enabled.
9. Click `Manage` on the Stripe method, then click the `Settings` tab.
* If live keys were entered, ensure `Enable test mode` is disabled.
* If test keys were entered, ensure `Enabled test mode` is enabled.
10. Click the `Edit account keys` button and ensure the key fields are populated (even after a refresh).

When adding testing and publishable keys that start with `pk_test_`, this does not mark setup as "complete" but I think this is open to interpretation and we can update accordingly.

#### Follow-up
I was not able to successfully connect Stripe when connecting to WordPress through a secured connection (HTTPS). So, please let me know if that approach also needs to be tested and I can take a second look 👀

lovo-h commented:
Also note, there were some `Linting / PHP linting` tests that failed. We may need to handle these before merging 😃

dechov commented:
The setup help text doesn't really apply when the OAuth button is shown – would it be possible to have it appear with the fields only? Or modify it to pertain to the connection flow?

<img width="731" src="https://user-images.githubusercontent.com/1867547/144536857-92cdca83-356c-41a8-a6e6-f09e304cee66.png">

lovo-h committed:
Merge branch 'develop' into add/remote-gateway-settings

lovo-h commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -1162,4 +1163,86 @@ public function validate_account_statement_descriptor_field( $param, $value, $ma
+	public function get_setup_help_text() {
+		return sprintf(
+			__( 'Your API details can be obtained from your <a href="%s">Stripe account</a>. Don’t have a Stripe account? <a href="%s">Create one.</a>', 'woocommerce-gateway-stripe' ),
+			'https://stripe.com/docs/keys',
```
Based on the context of the text, it appears that the intended URL for the first string should be `https://dashboard.stripe.com/apikeys`, as opposed to `https://stripe.com/docs/keys`.

Is this correct? Or, should the context of the text be updated to indicate that the link will navigate the user to the Stripe docs, instead?

lovo-h replied:
CC @joshuatf

joshuatf replied:
I believe that is correct, @lovo-h!

lovo-h committed:
PHPCS: Use short array syntax to define arrays

Before, the linter was failing the following PHPCS rule:
- "Short array syntax must be used to define arrays"

lovo-h committed:
PHPCS: Ignore sanitization for simple query parsing

Before, PHPCS would warn that $_SERVER['HTTP_REFERRER']
was not being sanitized correctly. But, because it is
being used for simple comparisons to check 'page' and
'task', the warning can be surpressed.

lovo-h committed:
PHPCS: Add translators helper comment & update API details link

Before, the Stripe API details link was referencing the Stripe docs
when it was intended to reference the Stripe account's API details. This
commit fixes that.

Also, fixing a misaligned line of code.

lovo-h committed:
Merge branch 'develop' into add/remote-gateway-settings

lovo-h commented:
> The setup help text doesn't really apply when the OAuth button is shown – would it be possible to have it appear with the fields only? [Comment reference](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1569#issuecomment-985172253).

I agree. The text seems inconsistent with the `Connect` button (see image below).

![image](https://user-images.githubusercontent.com/6844516/146592487-62bc1ced-938e-4e32-af4a-3ffd245f5811.png)

I am not sure whether we should just remove the text or replace it with some other contextual text. We may need to get Design involved. Thoughts?

For reference, [woocommerce-admin/client/tasks/fills/PaymentGatewaySuggestions/components/Setup/Configure.js:126](https://github.com/woocommerce/woocommerce-admin/blob/main/client/tasks/fills/PaymentGatewaySuggestions/components/Setup/Configure.js#L126) is the line in question.

**Edit**: The same thing can be said about the default state (see image below).

![image](https://user-images.githubusercontent.com/6844516/146597014-57bef429-499e-4492-8eb8-dce885a19e19.png)

Reference: [woocommerce-admin/client/tasks/fills/PaymentGatewaySuggestions/components/Setup/Configure.js:153](https://github.com/woocommerce/woocommerce-admin/blob/main/client/tasks/fills/PaymentGatewaySuggestions/components/Setup/Configure.js#L153).

cc @joshuatf

lovo-h committed:
Add comment to clarify the reason for short-circuiting early

Before, the if-then block wasn't 100% clear as to what was going on.
The added comment helps clarify what's going on in the code, a bit.

lovo-h commented:
The code LGTM 👍   
But since I made additional changes, I will need someone else to code review (cc @dechov).

Just to follow up on my [comments about updating the testing instructions](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1569#issuecomment-975999559). I don't think we need to update them since they tested as intended.

lovo-h committed:
Merge branch 'develop' into add/remote-gateway-settings

dechov commented:
> Make sure that the Stripe gateway is moved to the "Enabled payment gateways" section.

Just did a test, and it worked fine except that the Stripe gateway didn't get _enabled_ after redirect – doesn't look like this code change touches that setting?

lovo-h commented:
> Just did a test, and it worked fine except that the Stripe gateway didn't get enabled after redirect...

@dechov, I just re-ran the test and here was my result:

![](https://cdn-std.droplr.net/files/acc_1209167/iKw4aR)

Could you expand on what steps you took in order to test this? In the meantime, I will go ahead and update the testing instructions to make them a bit more clear 🙂 

> ... doesn't look like this code change touches that [enable] setting?

The gateway gets enabled from within `woocommerce-admin`: [See the following line for more details](https://github.com/woocommerce/woocommerce-admin/blob/d5cd7883a79a0c5586c82a94461df99285744eb4/client/tasks/fills/PaymentGatewaySuggestions/components/Setup/Configure.js#L65).

lovo-h commented:
**_This conversation was moved over here from Slack_**:

@rafaelzaleski, I am working on PR [#1569](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1569#issuecomment-1018864407) and I planned to update the Stripe extension so that it automatically gets enabled after the merchant has entered their Stripe account keys. The reason for this decision is to maintain consistent behavior with what the [WooCommerce Admin extension already does](https://github.com/woocommerce/woocommerce-admin/blob/d5cd7883a79a0c5586c82a94461df99285744eb4/client/tasks/fills/PaymentGatewaySuggestions/components/Setup/Configure.js#L65) when the merchant goes to `Set up payments` for Stripe.

But, if I recall correctly, there was a previous conversation in which it was decided that this should not occur. Hence, we [prevented the Stripe extension from being automatically enabled](https://github.com/woocommerce/woocommerce-gateway-stripe/commit/23e67927fb0d5f5a39fb8daa7034324b4a4380d5#diff-ab266b59ba60eb9c46e7df7368c30ad54be4e33b051703b02f5e54a62ba5bf1eL131). Could you refresh my memory as to the reason for this decision? And depending on the reasoning, should we advise that the WooCommerce Admin extension not automatically enable the Stripe payment gateway, once the merchant enters the Stripe account keys, as well?

cc @dechov 

---

@rafaelzaleski response:

> I discussed this matter very briefly with [@cesarcosta99] in a PR. We agreed that you could not assume the shop admin wants to start using Stripe right away just because they added the keys or activated the plugin.
> 
> From my field experience, sometimes, we would set up everything to run an experiment later. If changing the gateway would change conversion rates, for example.
>
> You can probably discuss this more with the design team. But if you enable the gateway for the user, you are assuming they want to start using Stripe right at the point where you do that.

lovo-h commented:
> You can probably discuss this more with the design team. But if you enable the gateway for the user, you are assuming they want to start using Stripe right at the point where you do that

@LevinMedia, would you be able to provide some guidance on how to move forward with how this functionality should work? More specifically, when a merchant connects using their Stripe account keys (or via the express connect), should we automatically enable the Stripe payment gateway ([see example](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1569#issuecomment-1018864407))? Or, should we leave it disabled and let the merchant manually enable it at a later time?

In the comment above, @rafaelzaleski brings up some [valid reasons for not automatically enabling the Stripe payment gateway](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1569#issuecomment-1021545479) but we wanted to verify this decision.

cesarcosta99 commented:
Just wanted to note that, in case we go ahead and change the extension to automatically enable the gateway after setup, we need to update the docs as well. It's currently mentioned over there that the gateway remains disabled after the initial setup.

lovo-h committed:
Code cleanup: Wrap long line into multiple lines

Also, replace 'else if' with 'elseif'.

lovo-h committed:
Enable Stripe payment method after connecting an account

Before, the Stripe payment method was not automatically enabled after
the merchant connected their Stripe account through "Connect" or by
manually entering Stripe account keys.

Now, the Stripe payment method is automatically enabled after the
merchant connects their Stripe account through either approach.

Fixes #2314

lovo-h committed:
Auto disable Stripe payment method once account keys are removed

Before, if the merchant removed their Stripe account keys, the Stripe
payment method would remain enabled. Now, it is automatically disabled
if the account keys are removed.

lovo-h committed:
Merge branch 'develop' into add/remote-gateway-settings

lovo-h committed:
Update changelog: Enable Stripe payment method after connecting

Also: Add onboarding payment gateway setup methods.

dechov commented on includes/admin/class-wc-rest-stripe-account-keys-controller.php:
```diff
@@ -197,6 +197,11 @@ public function set_account_keys( WP_REST_Request $request ) {
 					&& ! trim( $settings['secret_key'] )
 					&& ! trim( $settings['test_publishable_key'] )
 					&& ! trim( $settings['test_secret_key'] );
+		// If all new keys are empty, then account is being deleted. We should disable the payment gateway.
```
Unless this means the actual Stripe account is being deleted, maybe we should say "disconnected", or else "deleted from the plugin settings"?

lovo-h committed:
Replacing "deleted" with "disconnected" in comment, for accuracy

The word "deleted" implies that the Stripe account will be deleted when,
in fact, it will be "disconnected". The rewording should prevent any
confusion.

dechov reviewed (approved):
Looks good & tests well 🎉
